File: spring-web-java/src/main/java/spring/web/java/domain/model/entity/Role.java
package spring.web.java.domain.model.entity;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Role extends Base {

	@Id
	@GeneratedValue
	@Column(name = "role_id")
	private Long id;
	private String name;

	@OneToMany(mappedBy = "role", cascade = CascadeType.ALL)
	private List<RolePermission> rolePermissions = new ArrayList<>();

	public List<RolePermission> getRolePermissions() {
		return Collections.unmodifiableList(rolePermissions);
	}

	public static Role create(String name) {
		Role role = new Role();

		role.name = name;

		return role;
	}

	public void associatePermission(RolePermission rolePermission) {
		rolePermissions.add(rolePermission);
		rolePermission.setRole(this);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/model/entity/Role.kt
package spring.web.kotlin.domain.model.entity

import jakarta.persistence.*

@Entity
class Role protected constructor(
    name: String,
) : Base() {
    @Id
    @GeneratedValue
    @Column(name = "role_id")
    var id: Long? = null
        protected set

    var name: String = name
        protected set

    @OneToMany(mappedBy = "role", cascade = [(CascadeType.ALL)])
    private val _rolePermissions: MutableList<RolePermission> = mutableListOf()

    @get:Transient
    val rolePermissions: List<RolePermission>
        get() = _rolePermissions.toList()

    companion object {
        fun create(name: String, rolePermission: List<RolePermission>) =
            Role(name = name).apply { rolePermission.forEach { associatePermission(it) } }
    }

    fun associatePermission(rolePermission: RolePermission) {
        _rolePermissions.add(rolePermission)
        rolePermission.associateRole(this)
    }
}